README
NAME
    PYX::Optimization - PYX optimization Perl class.

SYNOPSIS
     use PYX::Optimization;

     my $obj = PYX::Optimization->new(%parameters);
     $obj->parse($pyx, $out);
     $obj->parse_file($pyx_file, $out);
     $obj->parse_handler($pyx_file_handler, $out);

METHODS
  "new"
     my $obj = PYX::Optimization->new(%parameters);

    Constructor.

    *       "output_encoding"

            Output encoding. Default value is 'utf-8'.

    *       "output_handler"

            Output handler. Default value is STDOUT.

    Returns instance of object.

  "parse"
     $obj->parse($pyx, $out);

    Optimize PYX defined by string $pyx. Output print to output handler. If
    $out not present, use 'output_handler'.

    Returns undef.

  "parse_file"
     $obj->parse_file($pux_file, $out);

    Optimize PYX defined by file $pyx_file. Output print to output handler.
    If $out not present, use 'output_handler'.

    Returns undef.

  "parse_handler"
     $obj->parse_handler($pyx_file_handler, $out);

    Optimize PYX defined by file handler $pyx_file_handler. Output print to
    output handler. If $out not present, use 'output_handler'.

    Returns undef.

ERRORS
     new():
             From Class::Utils::set_params():
                     Unknown parameter '%s'.

EXAMPLE1
     use strict;
     use warnings;

     use PYX::Optimization;

     # Content.
     my $pyx_to_optimize = <<'END';
     (element
     - data \n data
     )element
     _       comment
     (element
     -                                 \n foo
     )element
     END

     PYX::Optimization->new->parse($pyx_to_optimize);

     # Output:
     # (element
     # -data data
     # )element
     # _comment
     # (element
     # -foo
     # )element

EXAMPLE2
     use strict;
     use warnings;

     use PYX::Optimization;

     if (@ARGV < 1) {
             print STDERR "Usage: $0 pyx_file\n";
             exit 1;
     }
     my $pyx_file = $ARGV[0];

     PYX::Optimization->new->parse_file($pyx_file);

     # Output:
     # Usage: __SCRIPT__ pyx_file

DEPENDENCIES
    Class::Utils, Encode, Error::Pure, PYX, PYX::Parser, PYX::Utils.

SEE ALSO
    Task::PYX
        Install the PYX modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/PYX-Optimization>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2011-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.07



Changes
0.07

0.06 2023-10-19T10:27:25+02:00
 - Fix typo in doc.
 - Improve texts in doc.
 - Rename example files to better names.
 - Update copyright years.

0.05 2023-10-15T16:55:51+02:00
 - Fix constructor call in doc.
 - Fix Makefile.PL fails with no '.' in @INC.
 - Update Module::Install to 1.21 version.

0.04 2021-02-21T02:41:29+01:00
 - Add missing return to code. Add blank line after arguments.
 - Fix output encoding and add tests.
 - UTF-8 -> utf-8

0.03 2021-02-20T12:38:00+01:00
 - Add PYX code for data and comment, which are removed.
 - Add descriptions to test cases.
 - Add test for constructor.
 - Add tests for parse().
 - Add tests for parse_handler().
 - Rewrite implicit encode() to Encode::encode().
 - Update README.

0.02 2021-02-20T12:16:20+01:00
 - Add 'output_encoding' parameter and fix output encoding for default
   UTF-8.
 - Fix bugtracker.
 - Rewrite implicit decode() to PYX::Utils::decode().
 - Rewrite implicit encode() to PYX::Utils::encode().
 - Update copyright years.
 - Update examples/ directory.

0.01 2020-05-01T18:15:58+02:00
 - First version.


Optimization.pm
package PYX::Optimization;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Encode;
use Error::Pure qw(err);
use PYX qw(char comment);
use PYX::Parser;
use PYX::Utils;

our $VERSION = 0.07;

# Constructor.
sub new {
	my ($class, @params) = @_;
	my $self = bless {}, $class;

	# Output encoding.
	$self->{'output_encoding'} = 'utf-8';

	# Output handler.
	$self->{'output_handler'} = \*STDOUT;

	# Process params.
	set_params($self, @params);

	# PYX::Parser object.
	$self->{'pyx_parser'} = PYX::Parser->new(
		'output_encoding' => $self->{'output_encoding'},
		'output_handler' => $self->{'output_handler'},
		'output_rewrite' => 1,
		'callbacks' => {
			'data' => \&_data,
			'comment' => \&_comment,
		},
		'non_parser_options' => {
			'output_encoding' => $self->{'output_encoding'},
		},
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;

	$self->{'pyx_parser'}->parse($pyx, $out);

	return;
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file, $out) = @_;

	$self->{'pyx_parser'}->parse_file($file, $out);

	return;
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;

	$self->{'pyx_parser'}->parse_handler($input_file_handler, $out);

	return;
}

# Process data.
sub _data {
	my ($pyx_parser_obj, $data) = @_;

	my $tmp = PYX::Utils::encode($data);
	if ($tmp =~ /^[\s\n]*$/) {
		return;
	}

	# TODO Preserve?

	# White space on begin of data.
	$tmp =~ s/^[\s\n]*//s;

	# White space on end of data.
	$tmp =~ s/[\s\n]*$//s;

	# White space on middle of data.
	$tmp =~ s/[\s\n]+/\ /sg;

	$data = PYX::Utils::decode($tmp);
	my $out = $pyx_parser_obj->{'output_handler'};
	my $encoded_output = Encode::encode(
		$pyx_parser_obj->{'non_parser_options'}->{'output_encoding'},
		char($data),
	);
	print {$out} $encoded_output, "\n";

	return;
}

# Process comment.
sub _comment {
	my ($pyx_parser_obj, $comment) = @_;

	my $tmp = PYX::Utils::encode($comment);
	if ($tmp =~ /^[\s\n]*$/) {
		return;
	}
	$tmp =~ s/^[\s\n]*//s;
	$tmp =~ s/[\s\n]*$//s;
	$comment = PYX::Utils::decode($tmp);
	my $out = $pyx_parser_obj->{'output_handler'};
	my $encoded_output = Encode::encode(
		$pyx_parser_obj->{'non_parser_options'}->{'output_encoding'},
		comment($comment),
	);
	print {$out} $encoded_output, "\n";

	return;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::Optimization - PYX optimization Perl class.

=head1 SYNOPSIS

 use PYX::Optimization;

 my $obj = PYX::Optimization->new(%parameters);
 $obj->parse($pyx, $out);
 $obj->parse_file($pyx_file, $out);
 $obj->parse_handler($pyx_file_handler, $out);

=head1 METHODS

=head2 C<new>

 my $obj = PYX::Optimization->new(%parameters);

Constructor.

=over 8

=item * C<output_encoding>

Output encoding.
Default value is 'utf-8'.

=item * C<output_handler>

Output handler.
Default value is STDOUT.

=back

Returns instance of object.

=head2 C<parse>

 $obj->parse($pyx, $out);

Optimize PYX defined by string C<$pyx>.
Output print to output handler.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_file>

 $obj->parse_file($pux_file, $out);

Optimize PYX defined by file C<$pyx_file>.
Output print to output handler.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head2 C<parse_handler>

 $obj->parse_handler($pyx_file_handler, $out);

Optimize PYX defined by file handler C<$pyx_file_handler>.
Output print to output handler.
If C<$out> not present, use 'output_handler'.

Returns undef.

=head1 ERRORS

 new():
         From Class::Utils::set_params():
                 Unknown parameter '%s'.

=head1 EXAMPLE1

=for comment filename=remove_whitespace_from_comments.pl

 use strict;
 use warnings;

 use PYX::Optimization;

 # Content.
 my $pyx_to_optimize = <<'END';
 (element
 - data \n data
 )element
 _       comment
 (element
 -                                 \n foo
 )element
 END

 PYX::Optimization->new->parse($pyx_to_optimize);

 # Output:
 # (element
 # -data data
 # )element
 # _comment
 # (element
 # -foo
 # )element

=head1 EXAMPLE2

=for comment filename=run_on_pyx_file.pl

 use strict;
 use warnings;

 use PYX::Optimization;

 if (@ARGV < 1) {
         print STDERR "Usage: $0 pyx_file\n";
         exit 1;
 }
 my $pyx_file = $ARGV[0];

 PYX::Optimization->new->parse_file($pyx_file);

 # Output:
 # Usage: __SCRIPT__ pyx_file

=head1 DEPENDENCIES

L<Class::Utils>,
L<Encode>,
L<Error::Pure>,
L<PYX>,
L<PYX::Parser>,
L<PYX::Utils>.

=head1 SEE ALSO

=over

=item L<Task::PYX>

Install the PYX modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/PYX-Optimization>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2011-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.07

=cut


Optimalization.pm
package PYX::Optimalization;

use strict;
use warnings;

use Class::Utils qw(set_params);
use Error::Pure qw(err);
use PYX qw(char comment);
use PYX::Parser;
use PYX::Utils qw(encode decode);

our $VERSION = 0.01;

# Constructor.
sub new {
	my ($class, @params) = @_;
	my $self = bless {}, $class;

	# Output handler.
	$self->{'output_handler'} = \*STDOUT;

	# Process params.
	set_params($self, @params);

	# PYX::Parser object.
	$self->{'pyx_parser'} = PYX::Parser->new(
		'output_handler' => $self->{'output_handler'},
		'output_rewrite' => 1,
		'data' => \&_data,
		'comment' => \&_comment,
	);

	# Object.
	return $self;
}

# Parse pyx text or array of pyx text.
sub parse {
	my ($self, $pyx, $out) = @_;
	$self->{'pyx_parser'}->parse($pyx, $out);
}

# Parse file with pyx text.
sub parse_file {
	my ($self, $file) = @_;
	$self->{'pyx_parser'}->parse_file($file);
}

# Parse from handler.
sub parse_handler {
	my ($self, $input_file_handler, $out) = @_;
	$self->{'pyx_parser'}->parse_handler($input_file_handler, $out);
}

# Process data.
sub _data {
	my ($pyx_parser_obj, $data) = @_;
	my $tmp = encode($data);
	if ($tmp =~ /^[\s\n]*$/) {
		return;
	}

	# TODO Preserve?

	# White space on begin of data.
	$tmp =~ s/^[\s\n]*//s;

	# White space on end of data.
	$tmp =~ s/[\s\n]*$//s;

	# White space on middle of data.
	$tmp =~ s/[\s\n]+/\ /sg;

	$data = decode($tmp);
	my $out = $pyx_parser_obj->{'output_handler'};
	print {$out} char($data), "\n";
}

# Process comment.
sub _comment {
	my ($pyx_parser_obj, $comment) = @_;
	my $tmp = encode($comment);
	if ($tmp =~ /^[\s\n]*$/) {
		return;
	}
	$tmp =~ s/^[\s\n]*//s;
	$tmp =~ s/[\s\n]*$//s;
	$comment = decode($tmp);
	my $out = $pyx_parser_obj->{'output_handler'};
	print {$out} comment($comment), "\n";
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

PYX::Optimalization - TODO

=head1 SYNOPSIS

TODO

=head1 METHODS

=over 8

=item C<new(%parameters)>

 Constructor.

=over 8

=item * C<output_handler>

TODO

=back

=item C<parse()>

TODO

=item C<parse_file()>

TODO

=item C<parse_handler()>

TODO

=back

=head1 ERRORS

 Mine:
   TODO

 From Class::Utils::set_params():
   Unknown parameter '%s'.

=head1 EXAMPLE

 use strict;
 use warnings;

 use PYX::Optimalization;

 # PYX::Optimalization object.
 my $pyx = PYX::Optimalization->new(
   TODO
 );

=head1 DEPENDENCIES

L<Class::Utils>,
L<Error::Pure>,
L<PYX::Utils>.

=head1 SEE ALSO

TODO

=head1 AUTHOR

Michal Špaček L<skim@cpan.org>.

=head1 LICENSE AND COPYRIGHT

BSD license.

=head1 VERSION

0.01

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'PYX optimization Perl class.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'PYX-Optimization';
readme_from 'Optimization.pm';
recursive_author_tests('xt');
requires 'Class::Utils' => 0.05;
requires 'Error::Pure' => 0.15;
requires 'PYX' => 0.01;
requires 'PYX::Parser' => 0.08;
requires 'PYX::Utils' => 0.01;
requires 'perl' => '5.6.2';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/PYX-Optimization/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/PYX-Optimization';
resources 'repository' => 'git://github.com/michal-josef-spacek/PYX-Optimization';
test_requires 'English' => 0;
test_requires 'File::Object' => 0.07;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
test_requires 'Test::Pod' => 0;
test_requires 'Test::Pod::Coverage' => 0;
test_requires 'Unicode::UTF8' => 0;
tests_recursive;
version '0.07';

# Run.
WriteAll();
